### PR TITLE
Make servoMixer work for types other than CUSTOM_PLANE

### PR DESCRIPTION
**Pull request by lazd**
### Motivation

I wanted to use the servo mixer to assign a certain switch to control a servo on a certain ESC header on my quadcopter.

This change lets you enable the servo mixer for and create rules to do just that.

### Usage: Control servos on a QUADX

```
# Be a QUADX
mixer QUADX
 
# Enable servo mixer
feature SERVO_MIXER
 
# Reset servo mixer
smix reset
 
# Control servo output 1 with channel 7 (AUX_3)
smix 1 1 7 100 0 0 100 0
 
# Control servo output 2 with channel 8 (AUX_4)
smix 1 2 8 100 0 0 100 0
```

The wiring is as follows:

Header 1: Servo 1
Header 2: Servo 2
Header 3: Motor 1
Header 4: Motor 2
Header 5: Motor 3
Header 6: Motor 4

### Changes

* Added SERVO_MIXER feature to enable the servo mixer regardless of craft type
* Reset servos before applying servo values due to SERVO_TILT feature or GIMBAL type
* Servo mixer rules can override values set by SERVO_TILT feature and GIMBAL type
* Only apply rates to servo values affected by servo mixer rules
* Apply servo values to channels 1 and 2 when SERVO_MIXER feature is enabled
* Fix bug where input values in servoMixer were reversed due to subtracting the RC value from the RC middle value
* Center all servos by default
* Do gimbal AUX forwarding inside of writeServos()

## When is servo mixer enabled?

Assuming you've set `feature SERVOMIXER`, the servo mixer is enabled as follows, based on your type:

###  Servo mixer is enabled and works as it currently does, but only on the servos that are already used by this multitype

* MULTITYPE_CUSTOM_PLANE
* MULTITYPE_FLYING_WING
* MULTITYPE_AIRPLANE
* MULTITYPE_BI
* MULTITYPE_TRI
* MULTITYPE_DUALCOPTER
* MULTITYPE_SINGLECOPTER
* MULTITYPE_GIMBAL

### Servo mixer works for servos 1 and 2

* MULTITYPE_GIMBAL
* MULTITYPE_QUADP
* MULTITYPE_QUADX
* MULTITYPE_GIMBAL
* MULTITYPE_Y6
* MULTITYPE_HEX6
* MULTITYPE_Y4
* MULTITYPE_HEX6X
* MULTITYPE_OCTOX8
* MULTITYPE_OCTOFLATP
* MULTITYPE_OCTOFLATX
* MULTITYPE_HELI_120_CCPM
* MULTITYPE_HELI_90_DEG
* MULTITYPE_VTAIL4
* MULTITYPE_HEX6H
* MULTITYPE_PPM_TO_SERVO
* MULTITYPE_ATAIL4
* MULTITYPE_CUSTOM
* MULTITYPE_CUSTOM_PLANE

### CODE REVIEW

File: src/mixer.c [hunk 1]
```diff
@@ -463,6 +474,13 @@ void writeAllMotors(int16_t mc)
     writeMotors();
 }
 
+static void resetServos(void) {
```
Comment (trollcop) [inline src/mixer.c]: style

File: src/mixer.c [hunk 2]
```diff
@@ -511,24 +526,36 @@ static void servoMixer(void)
...
                 else if (currentOutput[i] > input[from])
                     currentOutput[i] = constrain(currentOutput[i] - currentServoMixer[i].speed, input[from], currentOutput[i]);
             }
 
-            servo[target] += servoDirection(target, from) * constrain(((int32_t)currentOutput[i] * currentServoMixer[i].rate) / 100, min, max);
-        } else
-            currentOutput[i] = 0;
-    }
+            // start with the output value
+            servo[target] = (int32_t)currentOutput[i];
```
Comment (trollcop) [inline src/mixer.c]: I'm pretty sure this should be either int or uint 16, not 32

File: src/mixer.c [hunk 3]
```diff
@@ -554,29 +581,20 @@ void mixTable(void)
...
-            servoMixer();
-            break;
-        case MULTITYPE_GIMBAL:
-            servo[0] = (((int32_t)cfg.servoConf[0].rate * angle[PITCH]) / 50) + servoMiddle(0);
-            servo[1] = (((int32_t)cfg.servoConf[1].rate * angle[ROLL]) / 50) + servoMiddle(1);
-            break;
+    if (core.useServo == 1) {
+        // reset all servos
+        resetServos();
     }
```
Comment (trollcop) [inline src/mixer.c]: above would look way better as
// reset all servos
if (core.useServo == 1)
    resetServos();

File: src/mixer.c [hunk 4]
```diff
@@ -588,20 +606,15 @@ void mixTable(void)
         }
     }
 
+    if (core.useServo == 1) {
+        // run the servo mixer if necessary
+        servoMixer();
+    }
+
     // constrain servos
```
Comment (trollcop) [inline src/mixer.c]: same on this one as previous comment

File: src/mixer.c [hunk 5]
```diff
@@ -463,6 +474,13 @@ void writeAllMotors(int16_t mc)
     writeMotors();
 }
 
+static void resetServos(void) {
+    // reset all servos to their middle value
+    uint8_t i;
+    for (i = 0; i < MAX_SERVOS; i++)
```
Comment (trollcop) [inline src/mixer.c]: space between variable definition and code

File: src/mixer.c (hunk 2 above)
Comment (lazd) [inline src/mixer.c]: The original code cast it as a uint_32, I merely refactored the expression so it was easier to maintain.

```c
servo[target] += servoDirection(target, from) * constrain(((int32_t)currentOutput[i] * currentServoMixer[i].rate) / 100, min, max);
```

File: src/mixer.c [hunk 6]
```diff
@@ -588,20 +607,15 @@ void mixTable(void)
         }
     }
 
+    // run the servo mixer if necessary
+    if (core.useServo == 1) {
+        servoMixer();
+    }
```
Comment (trollcop) [inline src/mixer.c]: single line if does not need braces

File: src/mixer.c [hunk 7]
```diff
@@ -554,29 +582,20 @@ void mixTable(void)
...
-            servoMixer();
-            break;
-        case MULTITYPE_GIMBAL:
-            servo[0] = (((int32_t)cfg.servoConf[0].rate * angle[PITCH]) / 50) + servoMiddle(0);
-            servo[1] = (((int32_t)cfg.servoConf[1].rate * angle[ROLL]) / 50) + servoMiddle(1);
-            break;
+    // reset all servos
+    if (core.useServo == 1) {
+        resetServos();
     }
```
Comment (trollcop) [inline src/mixer.c]: single line if does not need braces

File: src/mixer.c [hunk 8]
```diff
@@ -588,20 +606,14 @@ void mixTable(void)
         }
     }
 
+    // run the servo mixer if necessary
+    if (core.useServo == 1)
```
Comment (trollcop) [inline src/mixer.c]: just if (core.useServo) is fine, rest of the code does this

File: src/mixer.c [hunk 9]
```diff
@@ -463,6 +474,15 @@ void writeAllMotors(int16_t mc)
     writeMotors();
 }
 
+static void resetServos(void)
+{
+    uint8_t i;
```
Comment (trollcop) [inline src/mixer.c]: loop counter is faster as int

**Pull request merged**